### PR TITLE
test: add testcontainers

### PR DESCRIPTION
**Pull request by luky116**
### Purpose or design rationale of this PR

use testcontainer instead of start docker engine by cmd

https://github.com/scroll-tech/scroll/issues/436

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [ ] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [x] test: Adding missing tests or correcting existing tests

### Deployment tag versioning

Has `tag` in `common/version.go` been updated or have you added `bump-version` label to this PR?

- [x] No, this PR doesn't involve a new deployment, git tag, docker image tag
- [ ] Yes

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change
- [ ] Yes

### CODE REVIEW

File: common/database/db_test.go [hunk 1]
```diff
@@ -1,62 +1,27 @@
-package database
+package database_test
```
Comment (georgehao) [inline common/database/db_test.go:1]: why change this

File: common/testcontainers/testcontainers.go [hunk 2]
```diff
@@ -129,6 +134,30 @@ func (t *TestcontainerApps) GetL2GethEndPoint() (string, error) {
 	return endpoint, nil
 }
 
+// GetSqlxDBClient returns a sqlx.DB by connecting to the running postgres container
+func (t *TestcontainerApps) GetSqlxDBClient() (*sqlx.DB, error) {
+	endpoint, err := t.GetDBEndPoint()
+	if err != nil {
+		return nil, err
+	}
+	return sqlx.Open("postgres", endpoint)
```
Comment (georgehao) [inline common/testcontainers/testcontainers.go]: Oh, here is better switch to gorm

File: common/testcontainers/testcontainers_test.go [hunk 3]
```diff
@@ -3,16 +3,21 @@ package testcontainers
...
 
 // TestNewTestcontainerApps tests NewTestcontainerApps
 func TestNewTestcontainerApps(t *testing.T) {
 	var (
-		err      error
-		endpoint string
-		client   *ethclient.Client
+		err          error
+		endpoint     string
+		sqlDBclient  *sqlx.DB
```
Comment (georgehao) [inline common/testcontainers/testcontainers_test.go]: ditto

File: rollup/cmd/mock_app.go [hunk 4]
```diff
@@ -30,11 +28,11 @@ type MockApp struct {
 	args []string
 }
 
-// NewRollupApp TODO function will be replaced by NewRollupApp2
-func NewRollupApp(base *docker.App, file string) *MockApp {
-	rollupFile := fmt.Sprintf("/tmp/%d_rollup-config.json", base.Timestamp)
+// NewRollupApp return a new rollupApp manager, name mush be one them.
```
Comment (georgehao) [inline rollup/cmd/mock_app.go]: What this comment mean?

Comment (luky116) [reply rollup/cmd/mock_app.go]: This is the comment before. Do I need to modify it?

image

File: common/testcontainers/testcontainers.go (hunk 2 above)
Comment (luky116) [reply common/testcontainers/testcontainers.go]: done

File: common/testcontainers/testcontainers_test.go (hunk 3 above)
Comment (luky116) [reply common/testcontainers/testcontainers_test.go]: done

File: common/database/db_test.go (hunk 1 above)
Comment (luky116) [reply common/database/db_test.go:1]: image

package database  and database testcontainer  caused a circular reference, so i change the package name from database to database_test

File: rollup/cmd/mock_app.go (hunk 4 above)
Comment (georgehao) [reply rollup/cmd/mock_app.go]: yes

File: common/database/db_test.go (hunk 1 above)
Comment (georgehao) [reply common/database/db_test.go:1]: ok

File: rollup/cmd/mock_app.go (hunk 4 above)
Comment (luky116) [reply rollup/cmd/mock_app.go]: done

File: common/testcontainers/testcontainers.go [hunk 5]
```diff
@@ -6,6 +6,11 @@ import (
 	"log"
 	"time"
 
+	"scroll-tech/common/database"
+
+	"gorm.io/gorm"
+
+	_ "github.com/lib/pq" //nolint:golint
```
Comment (georgehao) [inline common/testcontainers/testcontainers.go]: i think this package don't need it

File: common/testcontainers/testcontainers.go [hunk 6]
```diff
@@ -6,6 +6,11 @@ import (
 	"log"
 	"time"
 
+	"scroll-tech/common/database"
```
Comment (georgehao) [inline common/testcontainers/testcontainers.go]: and this format is wrong

File: common/testcontainers/testcontainers_test.go [hunk 7]
```diff
@@ -3,16 +3,19 @@ package testcontainers
 import (
 	"testing"
 
+	"gorm.io/gorm"
```
Comment (georgehao) [inline common/testcontainers/testcontainers_test.go]: format

File: coordinator/cmd/api/app/mock_app.go [hunk 8]
```diff
@@ -10,10 +10,12 @@ import (
 	"testing"
 	"time"
 
-	"github.com/scroll-tech/go-ethereum/params"
+	"scroll-tech/common/testcontainers"
 
 	coordinatorConfig "scroll-tech/coordinator/internal/config"
 
+	"github.com/scroll-tech/go-ethereum/params"
```
Comment (georgehao) [inline coordinator/cmd/api/app/mock_app.go:13]: format

Comment (luky116) [reply coordinator/cmd/api/app/mock_app.go:13]: done

File: common/testcontainers/testcontainers_test.go (hunk 7 above)
Comment (luky116) [reply common/testcontainers/testcontainers_test.go]: done

File: common/testcontainers/testcontainers.go (hunk 5 above)
Comment (luky116) [reply common/testcontainers/testcontainers.go]: done

**Review by georgehao — DISMISSED** [review]

File: common/database/db_test.go [hunk 9]
```diff
@@ -1,62 +1,28 @@
...
-	gl.gethLogger = log.Root()
-
-	gl.Error(context.Background(), "test %s error:%v", "testError", errors.New("test error"))
-	gl.Warn(context.Background(), "test %s warn:%v", "testWarn", errors.New("test warn"))
-	gl.Info(context.Background(), "test %s warn:%v", "testInfo", errors.New("test info"))
-	gl.Trace(context.Background(), time.Now(), func() (string, int64) { return "test trace", 1 }, nil)
-}
-
 func TestDB(t *testing.T) {
+	var err error
```
Comment (colinlyguo) [inline common/database/db_test.go]: this var is not used?

**Review by colinlyguo — DISMISSED** [review]

File: common/database/db_test.go (hunk 9 above)
Comment (luky116) [reply common/database/db_test.go]: it is used
image

Comment (georgehao) [reply common/database/db_test.go]: can delete it

Comment (luky116) [reply common/database/db_test.go]: done

Comment (colinlyguo) [reply common/database/db_test.go]: resolved. btw, the err in the red circle is a "new" err, because the code uses `:=`.

**Review by colinlyguo — APPROVED** [review]
Can remove the original docker wrapper now?

**Comment by luky116** [comment]
> Can remove the original docker wrapper now?

I will remove it in next PR

**Review by georgehao — APPROVED** [review]

**Pull request merged**